packages/thirdweb/src/contract/deployment/utils/create-2-factory.ts

```diff
@@ -39,27 +40,74 @@ const SIGNATURE = {
 export async function computeCreate2FactoryAddress(
   options: ClientAndChain,
 ): Promise<string> {
-  // TODO add LRU cache
-  const commonFactory = getContract({
-    ...options,
-    address: COMMON_FACTORY_ADDRESS,
-  });
-  // if the common factory is deployed, return the address
-  if (await isContractDeployed(commonFactory)) {
+  const chainId = options.chain.id;
+
+  // special handling for chains with hardcoded gasPrice and gasLimit
+  if (CUSTOM_GAS_FOR_CHAIN[chainId]) {
+    const enforceEip155 = await isEIP155Enforced(options);
+    const eipChain = enforceEip155 ? chainId : 0;
+    const gasPrice = CUSTOM_GAS_FOR_CHAIN[eipChain.toString()]?.gasPrice;
+    const gasLimit = CUSTOM_GAS_FOR_CHAIN[eipChain.toString()]?.gasLimit;
+
+    const deploymentInfo = await _getCreate2FactoryDeploymentInfo(chainId, {
+      gasPrice,
+      gasLimit,
+    });
+
+    return deploymentInfo.predictedAddress;
+  }
+
+  // default flow
+  const allBinsInfo = await Promise.all([
+    // to generate EIP-155 transaction
+    ...CUSTOM_GAS_BINS.map((b) => {
+      return _getCreate2FactoryDeploymentInfo(chainId, { gasPrice: b });
+    }),
+
+    // to generate pre EIP-155 transaction, hence chainId 0
+    ...CUSTOM_GAS_BINS.map((b) => {
+      return _getCreate2FactoryDeploymentInfo(0, { gasPrice: b });
+    }),
+  ]);
+
+  const allFactories = await Promise.all(
+    allBinsInfo.map((b) => {
+      const tempFactory = getContract({
+        ...options,
+        address: b.predictedAddress,
+      });
+      return isContractDeployed(tempFactory);
+    }),
+  );
+
+  const indexOfCommonFactory = allBinsInfo.findIndex(
+    (b) => b.predictedAddress === COMMON_FACTORY_ADDRESS,
+  );
+  if (indexOfCommonFactory && allFactories[indexOfCommonFactory]) {
     return COMMON_FACTORY_ADDRESS;
   }
 
-  const enforceEip155 = await isEIP155Enforced(options);
-  const chainId = options.chain.id;
-  const custom = CUSTOM_GAS_FOR_CHAIN[chainId.toString()];
+  const indexOfExistingDeployment = allFactories.findIndex((b) => b);
+  if (
+    indexOfExistingDeployment &&
+    allBinsInfo &&
+    allBinsInfo[indexOfExistingDeployment]?.predictedAddress
+  ) {
+    // TODO: cleanup
+    return allBinsInfo[indexOfExistingDeployment]?.predictedAddress as string;
+  }
+
+  const [enforceEip155, gasPriceFetched] = await Promise.all([
+    isEIP155Enforced(options),
+    getGasPrice(options),
+  ]);
   const eipChain = enforceEip155 ? chainId : 0;
+  const bin =
+    CUSTOM_GAS_BINS.find((e) => e >= gasPriceFetched) || gasPriceFetched;
```

Comment:
Probably create a util function to do getNearestGasPriceBin? so we don't have to duplicate the algorithm across